vault/external_tests/seal_migration/seal_migration_test.go

```diff
@@ -0,0 +1,517 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	verifyBarrierConfig(t, b, wrapping.Transit, 1, 1, 1)
+	verifyBarrierConfig(t, r, wrapping.Shamir, keyShares, keyThreshold, 0)
+
+	cluster.EnsureCoresSealed(t)
+
+	return transitSeal
+}
+
+func unsealMigrate(t *testing.T, client *api.Client, keys [][]byte, transitServerAvailable bool) {
+
+	for i, key := range keys {
+
+		// Try to unseal with missing "migrate" parameter
+		_, err := client.Sys().UnsealWithOptions(&api.UnsealOpts{
+			Key: base64.StdEncoding.EncodeToString(key),
+		})
+		if err == nil {
+			t.Fatal("expected error due to lack of migrate parameter")
+		}
+
+		// Unseal with "migrate" parameter
+		resp, err := client.Sys().UnsealWithOptions(&api.UnsealOpts{
+			Key:     base64.StdEncoding.EncodeToString(key),
+			Migrate: true,
+		})
+
+		if i < keyThreshold-1 {
+			// Not enough keys have been provided yet.
+			if err != nil {
+				t.Fatal(err)
+			}
+		} else {
+			if transitServerAvailable {
+				// The transit server is running.
+				if err != nil {
+					t.Fatal(err)
+				}
+				if resp == nil || resp.Sealed {
+					t.Fatalf("expected unsealed state; got %#v", resp)
+				}
+			} else {
+				// The transit server is stopped.
+				if err == nil {
+					t.Fatal("expected error due to transit server being stopped.")
+				}
+			}
+			break
+		}
+	}
+}
+
+// verifyBarrierConfig verifies that a barrier configuration is correct.
+func verifyBarrierConfig(t *testing.T, cfg *vault.SealConfig, sealType string, shares, threshold, stored int) {
+	t.Helper()
+	if cfg.Type != sealType {
+		t.Fatalf("bad seal config: %#v, expected type=%q", cfg, sealType)
+	}
+	if cfg.SecretShares != shares {
+		t.Fatalf("bad seal config: %#v, expected SecretShares=%d", cfg, shares)
+	}
+	if cfg.SecretThreshold != threshold {
+		t.Fatalf("bad seal config: %#v, expected SecretThreshold=%d", cfg, threshold)
+	}
+	if cfg.StoredShares != stored {
+		t.Fatalf("bad seal config: %#v, expected StoredShares=%d", cfg, stored)
+	}
+}
+
+// initializeShamir initializes a brand new backend storage with Shamir.
+func initializeShamir(
+	t *testing.T, logger hclog.Logger,
+	storage teststorage.ReusableStorage, basePort int) (string, [][]byte) {
+
+	var baseClusterPort = basePort + 10
+
+	// Start the cluster
+	var conf = vault.CoreConfig{
+		Logger: logger.Named("initializeShamir"),
+	}
+	var opts = vault.TestClusterOptions{
+		HandlerFunc:           vaulthttp.Handler,
+		NumCores:              numTestCores,
+		BaseListenAddress:     fmt.Sprintf("127.0.0.1:%d", basePort),
+		BaseClusterListenPort: baseClusterPort,
+	}
+	storage.Setup(&conf, &opts)
+	cluster := vault.NewTestCluster(t, &conf, &opts)
+	cluster.Start()
+	defer func() {
+		storage.Cleanup(t, cluster)
+		cluster.Cleanup()
+	}()
+
+	leader := cluster.Cores[0]
+	client := leader.Client
+
+	// Unseal
+	if storage.IsRaft {
+		testhelpers.RaftClusterJoinNodes(t, cluster)
+		if err := testhelpers.VerifyRaftConfiguration(leader, numTestCores); err != nil {
```

Comment:
What I'd like to do then is keep `numTestCores`, with its setting of 5, and use it to set `opts.NumCores` as we are already doing.  But everywhere else in the tests I will use `len(cluster.Cores)`.  I believe this is in consonance with what you are proposing, while at the same time sticking with 5 for the number of test cores to prove that we haven't hardwired the default value of 3 anywhere in the tests.

Reply:
Is there a specific reason to use 5 vs 3 (the `vault.DefaultNumCores` value) for a multi-cluster setup?

Reply:
The specific reason is that I want to make sure I don't accidentally hard-wire the `ReusableStorage` stuff to use the default 3 cores, and since this is the only place right now in the code base that is using reusable storage, its the only place I can do that.

Reply:
Keeping in mind the `simplify use of numTestCores` commit (https://github.com/hashicorp/vault/pull/9085/commits/7271985fc6d797ba40e365c9a60475259039c399), which addresses some parts of Calvin's suggestion, I think the changes looks good.

The `numTestCores` variable as it is used now is only applicable to the seal migration package and is only used in places where the full cluster is not present. So, we can't do `len(cluster.Cores)`. Also, the reason quoted by Mike to have 5 cores seems valid as it is walking a case that ensures creation of necessary number of reusable storage artifacts.

Reply:
The only remaining usage of `numTestCores` seems to be on here:
https://github.com/hashicorp/vault/blob/7271985fc6d797ba40e365c9a60475259039c399/vault/external_tests/sealmigration/seal_migration_pre14_test.go#L80

My suggestion was not to use `len(cluster.Cores)` in that instance since `opts.NumCores` is the field that determines the number of cores the cluster should have, but rather re-use the const value, `vault.DefaultNumCores`, that we already have for multi-core cluster setups.